Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +95,329 @@ extension CertificatePolicy {
+            group.enter()
+            httpClient.post(url, body: requestData, headers: headers, options: options) { result in
+                defer { group.leave() }
+
+                switch result {
+                case .failure(let error):
+                    results.append(.failure(error))
+                case .success(let response):
+                    guard let responseData = response.body else {
+                        results.append(.failure(OCSPError.emptyResponseBody))
+                        return
+                    }
+
+                    let bytes = responseData.copyBytes()
+
+                    // Convert response to bio then OCSP response
+                    let bio = CCryptoBoringSSL_BIO_new(CCryptoBoringSSL_BIO_s_mem())
+                    defer { CCryptoBoringSSL_BIO_free(bio) }
+                    guard CCryptoBoringSSL_BIO_write(bio, bytes, numericCast(bytes.count)) > 0 else {
+                        results.append(.failure(OCSPError.responseConversionFailure))
+                        return
+                    }
+
+                    let response = d2i_OCSP_RESPONSE_bio(bio, nil)
+                    defer { OCSP_RESPONSE_free(response) }
+                    let basicResp = OCSP_response_get1_basic(response)
+                    defer { OCSP_BASICRESP_free(basicResp) }
+
+                    // This is just the OCSP response status, not the certificate's status
+                    guard OCSP_response_status(response) == OCSP_RESPONSE_STATUS_SUCCESSFUL,
+                        CCryptoBoringSSL_OBJ_obj2nid(response?.pointee.responseBytes.pointee.responseType) == NID_id_pkix_OCSP_basic,
+                        let basicRespData = basicResp?.pointee.tbsResponseData.pointee else {
+                        results.append(.failure(OCSPError.badResponse))
+                        return
+                    }
+
+                    // Inspect the OCSP response
+                    for i in 0 ..< sk_OCSP_SINGLERESP_num(basicRespData.responses) {
+                        guard let singleResp = sk_OCSP_SINGLERESP_value(basicRespData.responses, numericCast(i)),
+                            let certStatus = singleResp.pointee.certStatus else {
+                            results.append(.failure(OCSPError.badResponse))
+                            return
+                        }
+
+                        // Is the certificate in good status?
+                        let isCertGood = certStatus.pointee.type == V_OCSP_CERTSTATUS_GOOD
+                        results.append(.success(isCertGood))
+                        self.resultCache[cacheKey] = CacheValue(isCertGood: isCertGood, timestamp: DispatchTime.now())
+                        break
+                    }
+                }
+            }
+        }
+
+        group.notify(queue: callbackQueue) {
+            // If there's no result then something must have gone wrong
+            guard !results.isEmpty, results.compactMap({ $0.failure }).isEmpty else {
```

Comment:
A question here: this code as written fails closed in a number of scenarios, including if any of the OCSP responders are unavailable (or indeed all of them). Is that behaviour intentional? What should happen if all but one of the responders is unavailable but one returns success? What about if all of them are unavailable due to a CA outage: should users be unable to download new package collections or verify existing ones?

Reply:
I updated this logic to fail only when there is no result or if all results are failures. With the successful results, it fails if at least one of them indicates the cert is bad.

I guess a question to ask is if we want to base our decision on _all_ responders or a subset is acceptable. The implementation that Cory commented on requires all, but arguably that gives more weight to those who don't respond than the ones who do. The updated implementation uses whatever data it can (i.e., subset is ok).

If all responders are unavailable then the code wouldn't be able to verify the signature: new collection cannot be added (user can retry later or skip signature check if allowed); refresh of an existing collection would fail and its data would be removed from the database, but its URL will remain in the config for future refreshes.

Reply:
The most common approach would be to "fail open": that is, so long as one responder gave an authoritative response to follow that response, but if no responders are working then to assume the certificate is not revoked.

Reply:
OK makes sense. I will update the logic.